### PR TITLE
Use static pointers to frontend libraries

### PR DESCRIPTION
**Pull request by praasz**
### Details:
- Extend frontend libraries by static pointers.
- Add unit tests for frontend lib close
- Not use static FEM in ie network reader

### Tickets:
 - *82282*

### CODE REVIEW

File: src/frontends/common/src/manager.cpp [hunk 1]
```diff
@@ -20,16 +19,29 @@ class FrontEndManager::Impl {
...
     }
 
     ~Impl() = default;
 
+    FrontEnd::Ptr make_frontend(const ov::frontend::PluginInfo& plugin) {
+        auto fe_obj = std::make_shared<FrontEnd>();
+
+        fe_obj->m_shared_object = std::make_shared<FrontEndSharedData>(plugin.get_so_pointer());
+        fe_obj->m_actual = plugin.get_creator().m_creator();
+        m_shared_objects_map.emplace(plugin.get_creator().m_name, fe_obj->m_shared_object);
```
Comment (ilyachur) [inline src/frontends/common/src/manager.cpp:37]: It is not a thread safe approach

File: src/frontends/common/src/manager.cpp [hunk 2]
```diff
@@ -20,16 +19,29 @@ class FrontEndManager::Impl {
     std::mutex m_loading_mutex;
     std::vector<PluginInfo> m_plugins;
 
+    /// \brief map of shared object per frontend <frontend_name, frontend_so_ptr>
+    static std::map<std::string, std::shared_ptr<void>> m_shared_objects_map;
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/frontends/common/src/manager.cpp:23]: The `m_loading_mutex` is already used and this map is modified at same place where frontend is created from loaded plugin.

File: src/core/tests/frontend/onnx/CMakeLists.txt [hunk 3]
```diff
@@ -4,12 +4,12 @@
...
 
-file(GLOB_RECURSE SRC ${CMAKE_CURRENT_SOURCE_DIR}/*.cpp)
-list(FILTER SRC EXCLUDE REGEX standalone_build)
+file(GLOB SRC ${CMAKE_CURRENT_SOURCE_DIR}/*.cpp)
 
 add_executable(${TARGET_NAME} ${SRC})
 
-target_link_libraries(${TARGET_NAME} PRIVATE frontend_shared_test_classes openvino_onnx_frontend frontend_common)
+target_link_libraries(${TARGET_NAME}
+        PRIVATE frontend_shared_test_classes openvino_onnx_frontend frontend_common gtest_main_manifest)
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/tests/frontend/onnx/CMakeLists.txt:12]: I've done as was quiet simple to remove duplicates. But if you want can do another PR with this changes only.

File: src/frontends/common/src/manager.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (praasz) [reply src/frontends/common/src/manager.cpp]: There is also new mutex to guard this map.

File: src/frontends/common/include/openvino/frontend/manager.hpp [hunk 4]
```diff
@@ -75,6 +77,10 @@ class FRONTEND_API FrontEndManager final {
     FrontEnd::Ptr load_by_model_impl(const std::vector<ov::Any>& variants);
 
     std::unique_ptr<Impl> m_impl;
+
+    friend void ov::openvino_shutdown();
```
Comment (ilyachur) [inline src/frontends/common/include/openvino/frontend/manager.hpp:81]:
```suggestion
    friend FRONTEND_API void ov::openvino_shutdown();
```

File: src/frontends/common/include/openvino/frontend/manager.hpp [hunk 5]
```diff
@@ -13,6 +13,8 @@
 #include "openvino/frontend/visibility.hpp"
 
 namespace ov {
+// Forward declaration
+extern void openvino_shutdown();
```
Comment (ilyachur) [inline src/frontends/common/include/openvino/frontend/manager.hpp:17]:
```suggestion
```

**Comment by ilyachur** [comment]
Added label `do_not_merge` only in order to wait for feedback about architecture.

**Comment by ilyachur** [comment]
Could we extend OpenVINO bindings for `openvino_shutdown()` function?

File: src/inference/include/openvino/runtime/core.hpp [hunk 6]
```diff
@@ -672,4 +672,10 @@ class OPENVINO_RUNTIME_API Core {
     RemoteContext get_default_context(const std::string& device_name);
 };
 
+/**
+ * @brief Shut down the openvino by releasing dependent resources.
+ * @note This function should be used by advanced user to control unload the resources.
+ */
+OPENVINO_RUNTIME_API void openvino_shutdown();
```
Comment (ilyachur) [inline src/inference/include/openvino/runtime/core.hpp:679]:
```suggestion
OPENVINO_RUNTIME_API void shutdown();
```

File: src/inference/include/openvino/runtime/core.hpp [hunk 7]
```diff
@@ -672,4 +672,10 @@ class OPENVINO_RUNTIME_API Core {
     RemoteContext get_default_context(const std::string& device_name);
 };
 
+/**
+ * @brief Shut down the openvino by releasing dependent resources.
```
Comment (ilyachur) [inline src/inference/include/openvino/runtime/core.hpp:676]: Can we extend description why we need to have this function.
Example: https://github.com/protocolbuffers/protobuf/blob/520c601c99012101c816b6ccc89e8d6fc28fdbb8/src/google/protobuf/message_lite.h#L570

File: src/inference/include/openvino/runtime/core.hpp (hunk 6 above)
Comment (praasz) [reply src/inference/include/openvino/runtime/core.hpp:679]: Seems to be changed in #12293.

File: src/tests/functional/inference_engine/lib_close.cpp [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+                place = model->get_place_by_tensor_name(exp_name);
+            }
+
+            ov::shutdown();
+
+            ASSERT_NE(place, nullptr);
+            EXPECT_EQ(place->get_names().at(0), exp_name);
+        },
+        ".*");
+}
```
Comment (ilyachur) [inline src/tests/functional/inference_engine/lib_close.cpp:73]: Why do we have `EXPECT_DEATH` only for one test?

File: src/tests/functional/inference_engine/lib_close.cpp [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    Place::Ptr port_place;
+    {
+        auto fem = std::make_shared<ov::frontend::FrontEndManager>();
+        auto fe = fem->load_by_framework(frontend);
+        auto model = fe->load(model_path);
+        auto tensor_place = model->get_place_by_tensor_name(exp_name);
+        port_place = tensor_place->get_producing_port();
+    }
+
+    // ov::shutdown();
```
Comment (ilyachur) [inline src/tests/functional/inference_engine/lib_close.cpp:86]: Do we need this comment?

File: src/tests/functional/inference_engine/lib_close.cpp (hunk 8 above)
Comment (praasz) [reply src/tests/functional/inference_engine/lib_close.cpp:73]: This it additional test which should make segfault to check if ov::shutdown  close libs. 
Other check positive scenario if such think not occurs. I think one is enough.
But is should be renamed, it looks like I lose some local changes :(.

File: src/tests/functional/inference_engine/lib_close.cpp (hunk 9 above)
Comment (praasz) [reply src/tests/functional/inference_engine/lib_close.cpp:86]: No, garbage will remove.

File: src/frontends/tensorflow/tests/CMakeLists.txt [hunk 10]
```diff
@@ -28,39 +27,41 @@ target_compile_definitions(${TARGET_NAME} PRIVATE -D TEST_TENSORFLOW_MODELS_DIRN
...
-            COMMAND ${CMAKE_COMMAND}
-            -E cmake_echo_color --red "Warning: Unable to generate tensorflow test models. Running '${TARGET_NAME}' will likely fail"
-            )
-    add_custom_target(tensorflow_test_models DEPENDS unable_build_tensorflow_models.txt)
+        # Produce warning message at build time as well
+        add_custom_command(OUTPUT unable_build_tensorflow_models.txt
+                COMMAND ${CMAKE_COMMAND}
+                -E cmake_echo_color --red "Warning: Unable to generate tensorflow test models. Running '${TARGET_NAME}' will likely fail"
+        )
+        add_custom_target(tensorflow_test_models DEPENDS unable_build_tensorflow_models.txt)
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/frontends/tensorflow/tests/CMakeLists.txt:64]: Yes

File: src/frontends/tests/frontend/shared/include/utils.hpp [hunk 11]
```diff
@@ -65,4 +65,6 @@ inline bool exists(const std::string& file) {
 inline std::string make_model_path(const std::string& modelsRelativePath) {
     return CommonTestUtils::getModelFromTestModelZoo(modelsRelativePath);
 }
+
+std::string get_disabled_tests(const std::string& manifest_path);
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/frontends/tests/frontend/shared/include/utils.hpp:69]: Removed.

**Comment by ilyachur** [comment]
Do you investigate test fails?

**Comment by praasz** [comment]
> Do you investigate test fails?

Yes, model path issue solved. 
For static build lib test has to be skipped.

**Review by ilyachur — APPROVED** [review]

**Review by ilya-lavrenov — APPROVED** [review]

File: src/tests/functional/inference_engine/CMakeLists.txt [hunk 12]
```diff
@@ -69,18 +69,22 @@ addIeTargetTest(
         IE
 )
 
+target_compile_definitions(${TARGET_NAME} PRIVATE $<$<BOOL:${BUILD_SHARED_LIBS}>:"BUILD_SHARED_LIBS">)
```
Comment (itikhono) [inline src/tests/functional/inference_engine/CMakeLists.txt:72]: it looks like this line causes this issue on Ubuntu during cmake loading with this flag -DENABLE_TESTS:BOOL=ON

Compiler exited with error code 1: /usr/bin/c++
	<command-line>: error: macro names must be identifiers
	/tmp/compiler-file13789279785052064369:27:5: warning: "__clang__" is not defined, evaluates to 0 [-Wundef]
	   27 | #if __clang__
	      |     ^~~~~~~~~


cmd line:
cmake -DCMAKE_BUILD_TYPE=Debug -DENABLE_TESTS:BOOL=ON -DENABLE_HETERO:BOOL=OFF -DENABLE_PYTHON:BOOL=ON -DENABLE_MULTI:BOOL=OFF -DENABLE_ONEDNN_FOR_GPU:BOOL=OFF -DENABLE_CLDNN:BOOL=OFF -DENABLE_AUTO_BATCH:BOOL=OFF -DENABLE_AUTO:BOOL=OFF -DENABLE_INTEL_GPU:BOOL=OFF -DENABLE_INTEL_MYRIAD:BOOL=OFF -DENABLE_INTEL_MYRIAD_COMMON:BOOL=OFF -DENABLE_IR_V7_READER:BOOL=OFF -DENABLE_OV_PADDLE_FRONTEND:BOOL=OFF -S /home/itikhonov/OpenVINO/openvino -B /home/itikhonov/OpenVINO/openvino/cmake-build-debug

system info:
-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
OS: Ubuntu 20.04 or 22.04

Comment (praasz) [reply src/tests/functional/inference_engine/CMakeLists.txt:72]: I've check this cmake command on Ubuntu 20 compilers gcc 9.4, 11.1 (don't have 11.2) and same cmake version ( and one older version) no issue on cmake configuration step.
Is there any issue on CI check?
I can't reproduce it.

Comment (itikhono) [reply src/tests/functional/inference_engine/CMakeLists.txt:72]: for some reason it's not reproducible in CI, only locally but on 2 machines 
please check that you have BUILD_SHARED_LIBS=ON, but actually it should be ON by default

**Pull request merged**